app/src/components/main/withPagination/withPagination.jsx

```diff
@@ -0,0 +1,117 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Pagination } from '@reportportal/ui-kit';
+import { useSelector } from 'react-redux';
+import { useLayoutEffect, useState } from 'react';
+import classNames from 'classnames/bind';
+import {
+  DEFAULT_PAGE_SIZE,
+  DEFAULT_PAGINATION,
+  defaultPaginationSelector,
+  PAGE_KEY,
+  SIZE_KEY,
+} from 'controllers/pagination';
+import { connectRouter } from 'common/utils';
+import { getStorageItem, updateStorageItem } from 'common/utils/storageUtils';
+import { userIdSelector } from 'controllers/user';
+import styles from './withPagination.scss';
+
+const cx = classNames.bind(styles);
+
+const PAGE = 'Page';
+const PAGE_SIZE = 'PageSize';
+const normalizeValue = (value) => (value > 0 ? value : undefined);
+
+export const withPagination = ({
+  namespace,
+  pageSizeOptions,
+  paginationSelector = defaultPaginationSelector,
+  defaultPageSize = DEFAULT_PAGE_SIZE,
+} = {}) => (WrappedComponent) =>
+  connectRouter(
+    (query) => ({
+      page: query[PAGE_KEY] && normalizeValue(Number(query[PAGE_KEY])),
+      size: query[SIZE_KEY] && normalizeValue(Number(query[SIZE_KEY])),
+    }),
+    {
+      updatePagination: (page, size) => ({ [PAGE_KEY]: page, [SIZE_KEY]: size }),
+    },
+    {},
+  )(({ captions, updatePagination, page: queryPage, size: queryPageSize, ...props }) => {
+    const userId = useSelector(userIdSelector);
+    const { totalElements, totalPages } = useSelector(paginationSelector);
+    const [pageSize, setPageSize] = useState();
+    const [page, setPage] = useState();
+
+    const updateStorage = (value, propertyName) => {
+      if (namespace) {
+        updateStorageItem(`${userId}_settings`, {
+          [`${namespace}${propertyName}`]: value,
+        });
+      }
+    };
+
+    useLayoutEffect(() => {
+      const userSettings = getStorageItem(`${userId}_settings`) || {};
+
+      const hashPage =
+        queryPage ||
+        normalizeValue(Number(userSettings[`${namespace}${PAGE}`])) ||
+        DEFAULT_PAGINATION[PAGE_KEY];
+
+      const hashPageSize =
+        queryPageSize ||
+        normalizeValue(Number(userSettings[`${namespace}${PAGE_SIZE}`])) ||
+        defaultPageSize;
+
+      setPageSize(hashPageSize);
+      setPage(hashPage);
+      updatePagination(hashPage, hashPageSize);
+      updateStorage(hashPage, PAGE);
+      updateStorage(hashPageSize, PAGE_SIZE);
+    }, []);
+
+    const changePage = (newPage) => {
+      setPage(newPage);
+      updateStorage(newPage, PAGE);
+      updatePagination(newPage, pageSize);
+    };
+
+    const changePageSize = (newPageSize) => {
+      setPageSize(newPageSize);
+      updateStorage(newPageSize, PAGE_SIZE);
+      updatePagination(page, newPageSize);
+    };
+
+    return (
+      <div className={cx('with-pagination-container')}>
+        <WrappedComponent {...props} />
+        <div className={cx('with-pagination')}>
+          <Pagination
```

Comment:
What do you think about not embedding the `Pagination` component within this HOC?
It can be more flexible to render it separately where needed as there can be a rendering conditions (like [here](https://github.com/reportportal/service-ui/pull/3943/files#diff-f3cd87c0fa63ab64cbd1cd6a028d53e89e45471b95d0bebcbf7c12e82761400eR262)) or the needs to implement some lazy loading controls that will anyway require the access to at least `changePage` function.

Reply:
Done.